sapicore/utils/sweep.py

```diff
@@ -0,0 +1,259 @@
+
+from sklearn.model_selection import ParameterGrid
+
+
+class Sweep:
+    """Reads a `search_space` dictionary and generates appropriate parameter combinations.
+
+    Parameters
+    ----------
+    search_space: dict
+        Dictionaries of argument keys and their possible values (or scipy distributions and parameters),
+        grouped by top-level keys specifying the sweep strategy to be used with each.
+        Search spaces are typically stored in the `sweep` key  of a :class:`~engine.component.Component`
+        configuration dictionary.
+
+
+    Search space parameters are expected to be found under one of four top-level keys:
+
+    1. **Fixed**: The argument takes the same value across all branches. E.g., if
+    settings["fixed"]["resample"]=100, resample will always take the value 100.
+
+    2. **Zipped**: The values of the arguments listed are iterated over such that the Nth value of any argument is
+    matched with the Nth value of other arguments. Combinations are repeated as many times as it takes
+    to reach the number of combinations dictated by construction or by specification.
+    E.g., if settings["zipped"]["freq_band"] = [[5, 10], [30, 50]] and ... ["resample"] = [50, 100],
+    then `self.combinations` will have freq_band = [5, 10] with resample = 50 in half the branches,
+    and freq_band = [30, 50] with resample = 100 in the other half.
+
+    3. **Grid**: The values of the keys listed are iterated over such that any value combination is
+    equally represented in the resulting dictionary. Combinations are repeated as many times as it takes to reach
+    the number of combinations dictated by construction or by specification.
+    E.g., if settings["zipped"]["freq_band"] = [[5, 10], [30, 50]] and ...["resample"] = [50, 100],
+    then `self.combinations` will have four qualitatively distinct groups, with (freq_band, resample) taking
+    one of the following combined values: ([5, 10], 50), ([5, 10], 100), ([30, 50], 50), ([30, 50], 100).
+
+    4. **Random**: The values of the parameter keys listed are drawn from a :mod:`scipy.stats` distribution given
+    under the key "method", with the arguments "args". E.g., if settings["random"]["resample"]["method"]="uniform"
+    and ...["args"] = [x, y], then `resample` values will be drawn from `U(x, x+y)`.
+
+    Example
+    -------
+    Define a search space and generate a list of parameter combination dictionaries:
+
+        >>> s = Sweep(search_space={"fixed": {"a": 4}, "grid": {"b": [6, 7], "c": [8, 9]}}, num_combinations=4)
+        >>> print(s())
+        [{'a': 4, 'b': 6, 'c': 8}, {'a': 4, 'b': 6, 'c': 9}, {'a': 4, 'b': 7, 'c': 8}, {'a': 4, 'b': 7, 'c': 9}]
+
+    See Also
+    --------
+    `Scipy Distributions <https://docs.scipy.org/doc/scipy/reference/stats.html#continuous-distributions>`_
+        For a list of supported probability distributions and their arguments.
+
+    """
+
+    def __init__(self, search_space: dict, num_combinations: int = None):
```

Comment:
At a high level, I think the approach is good, but it's better to make things simpler instead of stuffing everything into one class/dict.

E.g. at the least I think instead of search_space this class should take the 4 kwargs of `fixed=None`, `zipped=None`, `grid=None`, and `random=None`. Or, preferable, have 4 classes that each takes different parameters. E.g. `FixedSweep` and then the users can specify a list of these sweep classes, each used for one or more different parameters to sweep. Because I don't think these different methods have to ever be combined, right!?

Then, it would be nice if we could specify the parameters in a call to a function to the class. E.g. you'd do something like this:

```py
class GridSweep(BaseSweep):
    def __init__(self, num_combinations=None, ...):
        super().__init__(...)
        self.num_combinations = num_combinations
        self._search_space = {}

    def add(self, **kwargs)
        self._search_space.update(kwargs)

    def __call__(self):
        if not_computed_yet:
            compute_comb
        return combinations
```

So then, you can use it in a much nicer way instead of stuffing everything into one init call:

```py
grid = GridSweep()
grid.add(a=[5, 7], b=[3, 9])
grid.add(c=[88, 12])
combinations = grid()
```

Reply:
I really like this. Currently, combining them doesn't change their behavior/order, though I can anticipate use cases where you'd want to (like counterbalancing across grid and zipped parameters). I don't intend to capture every complex use case, though, and compartmentalizing `GridSweep` etc. would greatly simplify the API.